mrmustard/lab_dev/circuit_components_utils.py

```diff
@@ -71,6 +70,28 @@ def __init__(
             name="Tr",
         )
 
+    def __custom_rrshift__(self, other: CircuitComponent | complex) -> CircuitComponent | complex:
+        r"""A custom ``>>`` operator for the ``TraceOut`` component.
+        It allows ``TraceOut`` to carry the method that processes ``other >> TraceOut``.
+        We know that the trace in Bargmann is a Gaussian integral, and in
+        Fock it's a trace (rather than an inner product with the identity).
+        So we write two shortcuts here, and ``__rrshift__`` will be called first if
+        present in the ``__rshift__`` method of the first object (``other`` here).
+        """
+        ket = other.wires.output.ket
+        bra = other.wires.output.bra
+        idx_zconj = [bra[m].indices[0] for m in self.wires.modes & bra.modes]
+        idx_z = [ket[m].indices[0] for m in self.wires.modes & ket.modes]
+        if not ket or not bra:
```

Comment:
If there's a possibility for ket or bra to not exist, what does bra[m] and ket[m] do in the lines above?

Reply:
Then they return a `Wires` object with no wires. You could try, you know.

Reply:
1. when you get subsets of wires using `.ket`, `.bra`, `.output` etc... and `__getitem__` (e.g. `wires[2,3,4]`) you get back a `Wires` object, regardless of how many wires it contains, so it can have zero.
2. Wires have a `__bool__` method that allows you to use the entire object as a bool which `False` if there are no wires, so `not bra` if `bra` is a Wires object is triggered when there are no wires on the bra side. So `not ket or not bra` is triggered when an object only has wires on one side, and we need to add the other side.

Reply:
Cool! Thanks.